Add Python Info to ARCHITECTURE.md (#1117)
glotzerlab/hoomd-blue: ARCHITECTURE.md

written in C++/HIP.  Priority is given to ease of use for users in Python even
at the cost of code complexity (within reason).

**Note**: Most internal functions, classes, and methods are documented for
developers (where they are not feel free to add documentation or request it).
Thus, this section will not go over individual functions or methods in detail
except where necessary.

_TODO_: Add the basic inheritance and features of HOOMD-blue Python
objects/operations.

### Base Data Model

#### Definitions

> - added: An object is associated with a simulation
> - attached: An object has its corresponding C++ class instantiated and is
>   connected to a simulation

Reviewer: ```suggestion
- added: An object is associated with a `Simulation`.
- attached: An object has its corresponding C++ class instantiated and is
  connected to a `Simulation`.
```

- unattached: An object's data resides in pure Python and may be or not be
  added.
- type parameter: An attribute that must be specified for all types or groups of

>   types of a set length

Reviewer: ```suggestion
  types of a set length.
```

In Python, many base classes exist to facilitate code reuse. This leads to very
lean user facing classes, but requires more fundamental understanding of the model
to address cases where customization is necessary, or changes are required. The
first aspect of the data model discussed is that of most HOOMD operations or
related classes.

- `Operation`
    - `_HOOMDBaseObject`
        - `_HOOMDGetSetAttrBase`

>         - `_DependencyRelation`

Reviewer: This should be inverted with `_DependencyRelation`,  `_HOOMDGetSetAttrBase` at the top, then `_HOOMDBaseObject` as a child and `Operation` as a child of that.

#### `_DependencyRelation`

`_DependencyRelation` helps define a dependent dependency relationship between
two objects. The class inherited by `_HOOMDBaseObject` to handle dependent

> relationships between operations in Python. This feature has yet to be used to

Reviewer: "Yet to be used" is now inaccurate with the recent PRs starting to use it. Describe what it is used for.

to enable custom attribute querying and setting when necessary. See
`hoomd/operation.py` for the source code and the sections on `TypeParameter`'s
and `ParameterDict`'s for more information.

#### `_HOOMDBaseObject`

`_HOOMDBaseObject` combines `_HOOMDGetSetAttrBase` and `_DependencyRelation` to
provide dependency handling, validated and processed attribute setting,
pickling, and pybind11 C++ class syncing in an automated and structured way.
Most methods unique to or customized from base classes revolve around allowing
`_param_dict` and `_typeparam_dict` to sync to and from C++ when attached and
not when unattached. See `hoomd/operation.py` for source code.

### Attribute Validation and Defaults

In Python, five fundamental classes exist for value validation, processing, and
syncing when attached: `hoomd.data.parameterdicts.ParameterDict`,
`hoomd.data.parameterdicts.TypeParameterDict`,
`hoomd.data.parameterdicts.AttachedTypeParameterDict`,
`hoomd.data.typeparam.TypeParameter`, and
`hoomd.data.syncedlist.SyncedList`.
These can be/are used by `_HOOMDBaseObject` subclasses as well as others.

#### `ParameterDict`

> `ParameterDict` provide a mapping (`dict`) interface from attribute names to

Reviewer: ```suggestion
`ParameterDict` provides a mapping (`dict`) interface from attribute names to
```

validated and processed values. Each instance of `ParameterDict` has its own
specification defining the validation logic for its keys. `_HOOMDBaseObject`

> subclasses automatically sync the `ParameterDict` instance in the `_param_dict`

Reviewer: Define "sync" in the glossary above.

attribute.

This class should be used to define all attributes shared with C++ member

> variables that are on a per-object basis (i.e. not per type). A good example of
> this class can be seen in many HOOMD classes, but a good example is

Reviewer: ```suggestion
variables that are on a per-object basis (i.e. not per type). Examples of
this can be seen in many HOOMD classes.  One good example is
```

`hoomd.md.update.ReversePerturbationFlow`.

> 

Reviewer: Say something more about how the syncing is achieved. Specifically, we need to mention what the C++ classes need to expose for the automatic syncing to work. Also need to mention somewhere (maybe here, maybe somewhere else) that the `attach` method is responsible for instantiating the C++ class.

#### `TypeParameterDict` and `AttachedTypeParameterDict`

These classes work together to define validated mappings from types or groups of
types to values. The type validation and processing logic is identical to that
used of `ParameterDict`, but on a per key basis. In addition, these classes
support advanced indexing features compared to standard Python `dict` instances.
The class also supports smart defaults. These features can come with some
complexity, so looking at the code with `hoomd.data.typeconverter`,
`hoomd.data.smart_default`, and `hoomd.data.parameterdicts`, should help.

The class is used with `TypeParameter` to specify quantities such as pair
potential `params` (see `hoomd/md/pair/pair.py`) and HPMC shape specs (see
`hoomd/hpmc/integrate.py`).

> 

Reviewer: As above, add some mention of how this syncs with C++ and the naming conventions needed to map between them.
